aggregator/src/test/java/org/jenkinsci/plugins/workflow/steps/input/InputStepRestartTest.java

```diff
@@ -24,43 +24,78 @@
+import static org.junit.Assert.*;
 import org.junit.Test;
 import org.junit.runners.model.Statement;
+import org.jvnet.hudson.test.Issue;
+import org.jvnet.hudson.test.recipes.LocalData;
 
 public class InputStepRestartTest extends SingleJobTestBase {
 
+    @Issue("JENKINS-25889")
     @Test public void restart() {
         story.addStep(new Statement() {
             @Override public void evaluate() throws Throwable {
                 p = jenkins().createProject(WorkflowJob.class, "demo");
                 p.setDefinition(new CpsFlowDefinition("input 'paused'"));
                 startBuilding();
-                waitForWorkflowToSuspend();
-                assertTrue(b.isBuilding());
+                story.j.waitForMessage("paused", b);
             }
         });
         story.addStep(new Statement() {
             @Override public void evaluate() throws Throwable {
                 rebuildContext(story.j);
-                assertThatWorkflowIsSuspended();
-                StepExecution.applyAll(InputStepExecution.class, new Function<InputStepExecution,Void>() {
-                    @Override public Void apply(InputStepExecution exec) {
-                        try {
-                            exec.doProceedEmpty();
-                        } catch (IOException x) {
-                            assert false : x;
-                        }
-                        return null;
-                    }
-                }).get();
+                proceed(b);
                 story.j.assertBuildStatusSuccess(story.j.waitForCompletion(b));
+                sanity(b);
+            }
+        });
+    }
+    
+    private void proceed(WorkflowRun b) throws Exception {
+        InputAction a = b.getAction(InputAction.class);
+        assertNotNull(a);
+        assertEquals(1, a.getExecutions().size());
+        story.j.submit(story.j.createWebClient().getPage(b, a.getUrlName()).getFormByName(a.getExecutions().get(0).getId()), "proceed");
+    }
+
+    private void sanity(WorkflowRun b) throws Exception {
+        List<PauseAction> pauses = new ArrayList<PauseAction>();
+        for (FlowNode n : new FlowGraphWalker(b.getExecution())) {
+            pauses.addAll(PauseAction.getPauseActions(n));
+        }
+        assertEquals(1, pauses.size());
+        assertFalse(pauses.get(0).isPaused());
+        String xml = FileUtils.readFileToString(new File(b.getRootDir(), "build.xml"));
+        assertFalse(xml, xml.contains(InputStepExecution.class.getName()));
+    }
+
+    @Issue("JENKINS-25889")
+    @LocalData // from 1.4.2
+    @Test public void oldFlow() {
+        story.addStep(new Statement() {
+            @Override public void evaluate() throws Throwable {
+                WorkflowJob p = jenkins().getItemByFullName("p", WorkflowJob.class);
+                assertNotNull(p);
+                WorkflowRun b = p.getLastBuild();
+                assertNotNull(b);
+                assertEquals(1, b.getNumber());
+                proceed(b);
```

Comment:
Trying to proceed on a build that was not started? Or perhaps the LocalData has an unfinished build?